src/commands/shell.ts

```diff
@@ -0,0 +1,120 @@
+import { Command } from "../command";
+import { start } from "repl";
+import { yellow, red, bold } from "cli-color";
+import * as admin from "firebase-admin";
+import { getFirebaseConfig } from "../functionsConfig";
+import { getAccessToken } from "../api";
+import { getRefreshToken } from "../auth";
+import { requireAuth } from "../requireAuth";
+import { FirebaseError } from "../error";
+import { createContext, runInContext, Context, Script } from "vm";
+import { readFileSync, existsSync } from "fs";
+import fetch, { Response } from "node-fetch";
+import { DocumentSnapshot, Firestore, QuerySnapshot } from "@google-cloud/firestore";
+import api = require("../api");
+
+async function runScript(sandbox: Context, scriptPath: string) {
+  if (!existsSync(scriptPath)) {
+    throw new FirebaseError(`Script file ${bold(scriptPath)} does not exist.`);
+  }
+
+  try {
+    const script = new Script(readFileSync(scriptPath, { encoding: "utf8" }), {
+      filename: scriptPath,
+    });
+    return await Promise.resolve(script.runInContext(sandbox));
+  } catch (e) {
+    // show the full stacktrace for the user's custom script
+    throw new FirebaseError(`Error running script: ${e.stack}`);
+  }
+}
+
+function runRepl(sandbox: Context) {
+  async function shellEval(command: string, context: any, filename: string, callback: any) {
+    try {
+      let result = await Promise.resolve(
+        // Wrap in an async function to allow top-level await.
+        runInContext(`(async function(){ return ${command}; })()`, sandbox)
+      );
+
+      if (result instanceof QuerySnapshot) {
+        result = result.docs.map((doc) => simpleDocSnapshot(doc));
+      } else if (result instanceof DocumentSnapshot) {
+        result = simpleDocSnapshot(result);
+      } else if (typeof result.val === "function") {
+        result = result.val();
+      } else if (result instanceof Response) {
+        if (result.headers.get("content-type")?.includes("application/json")) {
+          result = await result.json();
+        } else {
+          result = await result.text();
+        }
+      }
+
+      callback(null, result);
+    } catch (e) {
+      callback(e);
+    }
+  }
+
+  const replServer = start({
+    prompt: `${yellow("firebase")}${red(">")} `,
+    eval: shellEval,
+    // writer: replWriter,
+  });
+
+  return new Promise((resolve) => {
+    replServer.on("exit", resolve);
+    process.on("SIGINT", resolve);
```

Comment:
Does `SIGINT` cause `exit` to fire too?

Reply:
I...don't remember why I have this here but I remember it's necessary. I think it catches a weird state where you had to hit Ctrl+C multiple times or something.